miniapps/shifted_methods/dist_solver.hpp

```diff
@@ -0,0 +1,322 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_DIST_FUNCTION_HPP
+#define MFEM_DIST_FUNCTION_HPP
+
+#include "mfem.hpp"
+
+namespace mfem
+{
+
+class DistanceSolver
+{
+protected:
+   void ScalarDistToVector(ParGridFunction &dist_s, ParGridFunction &dist_v);
+
+public:
+   // 0 is nothing / 1 is the main solver / 2 is full (solver + precond).
+   int print_level = 0;
+
+   DistanceSolver() { }
+   virtual ~DistanceSolver() { }
+
+   virtual void ComputeScalarDistance(Coefficient &zero_level_set,
+                                      ParGridFunction &distance) = 0;
+
+   virtual void ComputeVectorDistance(Coefficient &zero_level_set,
+                                      ParGridFunction &distance);
+};
+
+// K. Crane et al:
+// Geodesics in Heat: A New Approach to Computing Distance Based on Heat Flow
+class HeatDistanceSolver : public DistanceSolver
+{
+public:
+   HeatDistanceSolver(double diff_coeff)
+      : DistanceSolver(), parameter_t(diff_coeff), smooth_steps(0),
+        diffuse_iter(1), transform(true), vis_glvis(false) { }
+
+   void ComputeScalarDistance(Coefficient &zero_level_set,
+                              ParGridFunction &distance);
+
+   int parameter_t, smooth_steps, diffuse_iter;
+   bool transform, vis_glvis;
+};
+
+class GradientCoefficient : public VectorCoefficient
```

Comment:
Technically this is a (negative) NormalizedGradientCoefficient. I would recommend changing the name to reflect that.